src/wasm/wasm-validator.cpp

```diff
@@ -974,9 +974,14 @@ static void validateMemory(Module& module, ValidationInfo& info) {
   if (curr.shared) info.shouldBeTrue(info.features & Feature::Atomics, "memory", "memory is shared, but atomics are disabled");
   for (auto& segment : curr.segments) {
     if (!info.shouldBeEqual(segment.offset->type, i32, segment.offset, "segment offset should be i32")) continue;
-    info.shouldBeTrue(checkOffset(segment.offset, segment.data.size(), module.memory.initial * Memory::kPageSize), segment.offset, "segment offset should be reasonable");
+    info.shouldBeTrue(checkOffset(segment.offset, segment.data.size(), curr.initial * Memory::kPageSize), segment.offset, "segment offset should be reasonable");
     Index size = segment.data.size();
-    info.shouldBeTrue(size <= curr.initial * Memory::kPageSize, segment.data.size(), "segment size should fit in memory (initial)");
+    // If the memory is imported we don't actually know its initial size.
+    // Specifically wasm dll's import a zero sized memory which is perfectly
```

Comment:
To make sure I understand, is it that they import a memory with initial size 0 but maximum size big enough for the segments? If so maybe the comment could say "an initially zero-sized memory"?

Reply:
In an import statement the limits express a constraint. The memory must be at least `min` pages, and at most `max` pages. You could check here that a data segment was always less than the max size, but that's an instantiation error, not an validation error.